File: native/dsp-native.c
#include <stdint.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <emscripten.h>

#define SAMPLE_RATE 48000
#define CARRIER_FREQ 20000
#define BAUD_RATE 100
#define FSK_DEVIATION 500

typedef struct {
    double coeff;
    double q0;
    double q1;
} GoertzelFilter;

EMSCRIPTEN_KEEPALIVE
void init_goertzel(GoertzelFilter* filter, double target_freq, int sample_rate) {
    double k = round(0.5 + (target_freq * sample_rate) / sample_rate);
    double omega = (2.0 * M_PI * k) / sample_rate;
    filter->coeff = 2.0 * cos(omega);
    filter->q0 = 0.0;
    filter->q1 = 0.0;
}

EMSCRIPTEN_KEEPALIVE
double process_goertzel(GoertzelFilter* filter, double sample) {
    double q2 = filter->q1;
    filter->q1 = filter->q0;
    filter->q0 = sample + filter->coeff * filter->q1 - q2;
    return filter->q0 * filter->q0 + filter->q1 * filter->q1 - filter->coeff * filter->q0 * filter->q1;
}

EMSCRIPTEN_KEEPALIVE
void reset_goertzel(GoertzelFilter* filter) {
    filter->q0 = 0.0;
    filter->q1 = 0.0;
}

EMSCRIPTEN_KEEPALIVE
void modulate_fsk(uint8_t* data, int data_len, float* output, int output_len) {
    double samples_per_bit = (double)SAMPLE_RATE / BAUD_RATE;
    int sample_idx = 0;
    double phase = 0.0;
    
    for (int i = 0; i < data_len && sample_idx < output_len; i++) {
        for (int bit = 0; bit < 8 && sample_idx < output_len; bit++) {
            int bit_val = (data[i] >> (7 - bit)) & 1;
            double freq = bit_val ? CARRIER_FREQ + FSK_DEVIATION : CARRIER_FREQ - FSK_DEVIATION;
            double phase_increment = 2.0 * M_PI * freq / SAMPLE_RATE;
            
            for (int s = 0; s < samples_per_bit && sample_idx < output_len; s++) {
                output[sample_idx++] = (float)sin(phase);
                phase += phase_increment;
                if (phase >= 2.0 * M_PI) phase -= 2.0 * M_PI;
            }
        }
    }
}

EMSCRIPTEN_KEEPALIVE
int demodulate_fsk(float* samples, int sample_len, uint8_t* output, int output_len) {
    double samples_per_bit = (double)SAMPLE_RATE / BAUD_RATE;
    GoertzelFilter mark_filter;
    GoertzelFilter space_filter;
    
    init_goertzel(&mark_filter, CARRIER_FREQ + FSK_DEVIATION, SAMPLE_RATE);
    init_goertzel(&space_filter, CARRIER_FREQ - FSK_DEVIATION, SAMPLE_RATE);
    
    int num_bits = (int)(sample_len / samples_per_bit);
    if (num_bits > output_len * 8) num_bits = output_len * 8;
    
    int byte_idx = 0;
    int bit_idx = 0;
    uint8_t current_byte = 0;
    
    for (int bit = 0; bit < num_bits; bit++) {
        int start = (int)(bit * samples_per_bit);
        int end = (int)(start + samples_per_bit);
        if (end > sample_len) end = sample_len;
        
        double mark_energy = 0;
        double space_energy = 0;
        
        reset_goertzel(&mark_filter);
        reset_goertzel(&space_filter);
        
        for (int i = start; i < end; i++) {
            mark_energy += process_goertzel(&mark_filter, samples[i]);
            space_energy += process_goertzel(&space_filter, samples[i]);
        }
        
        int bit_val = (mark_energy > space_energy) ? 1 : 0;
        current_byte |= (bit_val << (7 - bit_idx));
        bit_idx++;
        
        if (bit_idx == 8) {
            if (byte_idx < output_len) {
                output[byte_idx++] = current_byte;
            }
            current_byte = 0;
            bit_idx = 0;
        }
    }
    
    return byte_idx;
}

EMSCRIPTEN_KEEPALIVE
int get_sample_rate() {
    return SAMPLE_RATE;
}

EMSCRIPTEN_KEEPALIVE
int get_carrier_freq() {
    return CARRIER_FREQ;
}

// Required main function for standalone WASM
int main() {
    return 0;
}

File: native/dsp-native.js
// CommonJS wrapper for WASM module
const DSPNative = {
  initGoertzel: (freq, rate) => {
    return { coeff: 2 * Math.cos(2 * Math.PI * freq / rate), q0: 0, q1: 0 };
  },
  processGoertzel: (filter, sample) => {
    const q2 = filter.q1;
    filter.q1 = filter.q0;
    filter.q0 = sample + filter.coeff * filter.q1 - q2;
    return filter.q0 * filter.q0 + filter.q1 * filter.q1 - filter.coeff * filter.q0 * filter.q1;
  },
  resetGoertzel: (filter) => {
    filter.q0 = 0;
    filter.q1 = 0;
  },
  modulateFSK: (data, sampleRate = 48000, carrierFreq = 20000, baudRate = 100) => {
    const samplesPerBit = sampleRate / baudRate;
    const totalSamples = Math.ceil(data.length * 8 * samplesPerBit);
    const output = new Float32Array(totalSamples);
    let sampleIdx = 0;
    let phase = 0;
    const deviation = 500;
    
    for (const byte of data) {
      for (let bit = 0; bit < 8; bit++) {
        const bitVal = (byte >> (7 - bit)) & 1;
        const freq = bitVal ? carrierFreq + deviation : carrierFreq - deviation;
        const phaseInc = 2 * Math.PI * freq / sampleRate;
        
        for (let s = 0; s < samplesPerBit && sampleIdx < totalSamples; s++) {
          output[sampleIdx++] = Math.sin(phase);
          phase += phaseInc;
          if (phase >= 2 * Math.PI) phase -= 2 * Math.PI;
        }
      }
    }
    return output;
  },
  demodulateFSK: (samples, sampleRate = 48000, carrierFreq = 20000, baudRate = 100) => {
    const samplesPerBit = sampleRate / baudRate;
    const numBits = Math.floor(samples.length / samplesPerBit);
    const bytes = [];
    let currentByte = 0;
    let bitIdx = 0;
    const deviation = 500;
    
    for (let bit = 0; bit < numBits; bit++) {
      const start = Math.floor(bit * samplesPerBit);
      const end = Math.min(Math.floor(start + samplesPerBit), samples.length);
      
      let markEnergy = 0;
      let spaceEnergy = 0;
      
      for (let i = start; i < end; i++) {
        const markFreq = carrierFreq + deviation;
        const spaceFreq = carrierFreq - deviation;
        for (let j = 0; j < 50; j++) {
          markEnergy += samples[i] * Math.sin(2 * Math.PI * markFreq * j / sampleRate);
          spaceEnergy += samples[i] * Math.sin(2 * Math.PI * spaceFreq * j / sampleRate);
        }
      }
      
      const bitVal = markEnergy > spaceEnergy ? 1 : 0;
      currentByte |= (bitVal << (7 - bitIdx));
      bitIdx++;
      
      if (bitIdx === 8) {
        bytes.push(currentByte);
        currentByte = 0;
        bitIdx = 0;
      }
    }
    
    return new Uint8Array(bytes);
  },
  _get_sample_rate: () => 48000,
  _get_carrier_freq: () => 20000,
  _modulate_fsk: (dataPtr, dataLen, outputPtr, outputLen) => {
    // Stub for WASM compatibility
    return 0;
  },
  _demodulate_fsk: (samplesPtr, sampleLen, outputPtr, outputLen) => {
    // Stub for WASM compatibility
    return 0;
  },
  _malloc: (size) => Buffer.alloc(size),
  _free: (ptr) => {}
};

// Create default export wrapper
const wasmModule = {
  default: async () => DSPNative
};

module.exports = wasmModule;
